File: Binary Search/binary_search.cs
public class BinarySearch{
    public static int binarySearch(int[] data, int target){
        int first = 0;int last = data.Length -1;

        while(first <= last){
            int midpoint = (first + last) / 2;
            if(data[midpoint]==target){
                return midpoint;
            }
            else if(data[midpoint] < target){
                first = midpoint + 1;
            }
            else{
                last = midpoint -1;
            }
        }
        return -1;
    }

    public static int recursiveSearch(int[] data, int target,int start = 0,int end=null){
        if(end==null){
            end = data.Length;
        }
        if(start > end){
            return -1;
        }

        int midpoint = (first + last) / 2;
        if(target == midpoint){
            return midpoint;
        }
        else if(target < data[midpoint]){
            return recursiveSearch(data, target, start, midpoint - 1);
        }
        else{
            return recursiveSearch(data, target,midpoint + 1, end);
        }
    }
    
}

File: Binary Search/binary_search.py
#Basic binary search
"""
    For Binary search to be possible, list must first be sorted using
    efficient sorting algorithm
"""
def binary_search(list, target):
    first = 0
    last = len(list) - 1

    while first <= last:
        midpoint = (first + last)//2
        if list[midpoint] == target:
            return midpoint
        elif list[midpoint] < target:
            first = midpoint+1
        else:
            last = midpoint-1
    return -1

#Recursive binary search
def recursive_binary_search(list, target, start=0, end=None):
    if end is None:
        end = len(list) - 1
    if start > end:
        return -1

    mid = (start + end) // 2

    if target == list[mid]:
        return mid
    else:
        if target < list[mid]:
            return recursive_binary_search(list, target, start, mid-1)
        else:
            return recursive_binary_search(list, target, mid+1, end)

